src/python/pants/engine/internals/build_files_test.py

```diff
@@ -658,6 +658,70 @@ def test_default_parametrized_groups(target_adaptor_rule_runner: RuleRunner) ->
     )
 
 
+def test_default_parametrized_groups_with_parametrizations(
```

Comment:
In my very vague understanding there's some differences between the `all` and the per-target-type code paths for `__defaults__` (but I realise now that I don't actually know how these work 😅 )... is it's worth also testing a `__defaults__({mock_tgt: ...})` version?

Reply:
Nah.. the differences here are well enough tested in the defaults tests suite.
What `all` does is equiv. to doing `__defaults__({(tgt_a, tgt_b, tgt_c, ...): ...})` for all defined target types, while also filtering the provided field values to only include fields that apply for the particular target type. And it is this field filtering that breaks when using nested parametrize with `all`, as that hides the fields from the filtering logic, resulting in potentially invalid field values passed to a target from the defaults.

Reply:
Ah okay. How does the failure surface to a user who does try to use this with `all`?

Reply:
It's detailed in this https://github.com/pantsbuild/pants/pull/21027#issuecomment-2155593758 👍🏽

Reply:
This also relates to #20925 that the applied field filtering for `all` only looks at the field name, not the type of value, so any targets that shares a field name but with different types will have issues when used with `all` (which is the topic of the linked issue, by you ;) )

Reply:
> It's detailed in this https://github.com/pantsbuild/pants/pull/21027#issuecomment-2155593758 👍🏽

Ah sorry! I see. That doesn't seem like a great message (e.g. the `__default__` defining the field may be very far from the target with the problem)... I don't think it should block this PR, but is there any quick wins for improving the diagnostics?

Or maybe this is an existing problem (as you say) so we can just ignore it for now...

Reply:
It's semi-new. I think it may surface already with the parametrization group, in which case this doesn't make it worse, at least.

Would prefer to tackle it in another PR regardless.